src/box/txn_limbo.c

```diff
@@ -53,6 +55,9 @@ txn_limbo_create(struct txn_limbo *limbo)
 	limbo->promote_greatest_term = 0;
 	latch_create(&limbo->promote_latch);
 	limbo->confirmed_lsn = 0;
+	limbo->confirmed_entry =
+		rlist_entry(limbo->queue, struct txn_limbo_entry, in_queue);
+	limbo->confirmed_entry_is_valid = false;
```

Comment:
Is this flag needed? Why can't the entry being `NULL` be counted as "invalid"?

Reply:
`confirmed_lsn` is not the same as `confirmed_entry->lsn`. 
There is only such a statement `confirmed_lsn >= confirmed_entry->lsn`. Perhaps `confirm_entry` is not the best name for it. May be it should be named as `confirmation_pointer` or something like.

`confirmed_entry_is_valid = true` means that the next transaction after `confirmed_entry` is synchronous and `ack_count` stores the correct value. Only in this case confirmed_entry is considered valid. But even if `confirmed_node` is not valid, it still stores a useful value. In this case, `confirmed_entry` stores an entry from which it will be possible to continue searching for the next synchronous transaction.

This was done this way only so as not to once again go through the entries that we had already checked. Let's imagine that `confirmed_entry` is literally the last confirmed synchronous transaction (`confirmed_lsn = confirmed_entry->lsn`). Then the following situation is possible:
1. limbo queue: [..., sync, async, async, async, async, ...] (all transactions after `confirmed_entry` are asynchronous)
`confirmed_entry` --^
Soon `CONFIRM` will be written to WAL and this synchronous transaction, along with all subsequent asynchronous transactions, will be deleted from limbo. But until this happens, they continue to hang in limbo.
2. We receive an ACK for the last transaction in the queue from one of the replicas in the replicaset.
3. We move on, starting with `confirmed_entry` in hopes of finding a confirmed synchronous transaction and updating `confirmed_lsn`. But we don’t find it.
https://github.com/Astronomax/tarantool/blob/6eee079a537a40c9721f6ca8a9fe616748f3b5c6/src/box/txn_limbo.c#L680-L684
4. But we don’t remember where we came, leave the `confirmed_entry` value the same.
5. Having received the same ACK from another replica, we will go through the same transactions again. And that's the problem.

There is no difference in performance on the `1mops_write` benchmark. But if you use synchronous and asynchronous replication together, this can be noticeable if there are noticeably more asynchronous transactions than synchronous ones.

Reply:
Ah, ok, I see, this clears many things out, thanks 🙏.